### PR TITLE
[ValueTracking] Fix assertion failure when `computeKnownFPClass` returns fcNone

### PR DESCRIPTION
**Pull request by dtcxzyw**
Fixes https://github.com/llvm/llvm-project/pull/92084#issuecomment-2114083188.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -1126,7 +1126,8 @@ static void computeKnownBitsFromOperator(const Operator *I,
       KnownFPClass Result = computeKnownFPClass(V, fcAllFlags, Depth + 1, Q);
       FPClassTest FPClasses = Result.KnownFPClasses;
 
-      if (Result.isKnownNever(fcNormal | fcSubnormal | fcNan)) {
+      if (FPClasses != fcNone &&
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:1129]: Can we just assume the none case is a 0?

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:1129]: I think comparing with fcNone explicitly is better.

See https://github.com/llvm/llvm-project/blob/8389177710c2cb7a7e0cf5c6c9b34aabb8048f39/llvm/lib/Analysis/ValueTracking.cpp#L5007-L5008
https://github.com/llvm/llvm-project/blob/8389177710c2cb7a7e0cf5c6c9b34aabb8048f39/llvm/lib/Analysis/ValueTracking.cpp#L5174-L5178
https://github.com/llvm/llvm-project/blob/8389177710c2cb7a7e0cf5c6c9b34aabb8048f39/llvm/lib/Analysis/ValueTracking.cpp#L4199

Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:1129]: No, I don't mean in terms of enum. I mean treat it as-if the known bits are a literal 0

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp]: Done.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -1126,8 +1126,13 @@ static void computeKnownBitsFromOperator(const Operator *I,
       KnownFPClass Result = computeKnownFPClass(V, fcAllFlags, Depth + 1, Q);
       FPClassTest FPClasses = Result.KnownFPClasses;
 
-      if (FPClasses != fcNone &&
-          Result.isKnownNever(fcNormal | fcSubnormal | fcNan)) {
+      // Treat it as zero if the use of I is unreachable.
+      if (FPClasses == fcNone) {
+        Known.setAllZero();
+        break;
+      }
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:1133]: Probably should do this as a separate PR, this probably has the need-to-consistently-be-zero-for-undef problem

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:1133]: I will leave it as a todo.

Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:1133]: It's probably OK as long as we handle undef correctly inside computeKnownFPClass, the contradictory paths implies UB happened somewhere

**Review by arsenm — APPROVED** [review]
